File: codes/r/www.codewars/www.002.count_vowels.r
# Source:
#     https://www.codewars.com/kata/54ff3102c1bad923760001f3
# Title:
#     Vowel Count
# Description:
#     Return the number (count) of vowels in the given string.
#     We will consider a, e, i, o, u as vowels for this Kata (but not y).
#     The input string will only consist of lower case letters and/or spaces.

countVowelsV1 <- function(aString) Reduce(function(result, letter) (if (letter %in% list("a", "i", "u", "e", "o", "A", "I", "U", "E", "O")) result + 1 else result), strsplit(aString, split = "")[[1]], 0)
cat(paste(sep = "", countVowelsV1("Hello World"), "\n")) # 3

countVowelsV2 <- function(aString) Reduce(function(result, letter) (if (toupper(letter) %in% list("A", "I", "U", "E", "O")) result + 1 else result), strsplit(aString, split = "")[[1]], 0)
cat(paste(sep = "", countVowelsV2("Hello World"), "\n")) # 3

countVowelsV3 <- function(aString) Reduce(function(result, letter) (if (grepl(letter, "aiueoAIUEO")) result + 1 else result), strsplit(aString, split = "")[[1]], 0)
cat(paste(sep = "", countVowelsV3("Hello World"), "\n")) # 3

countVowelsV4 <- function(aString) Reduce(function(result, letter) (if (grepl(toupper(letter), "AIUEO")) result + 1 else result), strsplit(aString, split = "")[[1]], 0)
cat(paste(sep = "", countVowelsV4("Hello World"), "\n")) # 3

countVowelsV5 <- function(aString) length(Filter(function(letter) (letter %in% list("a", "i", "u", "e", "o", "A", "I", "U", "E", "O")), strsplit(aString, split = "")[[1]]))
cat(paste(sep = "", countVowelsV5("Hello World"), "\n")) # 3

countVowelsV6 <- function(aString) length(Filter(function(letter) (toupper(letter) %in% list("A", "I", "U", "E", "O")), strsplit(aString, split = "")[[1]]))
cat(paste(sep = "", countVowelsV6("Hello World"), "\n")) # 3

countVowelsV7 <- function(aString) length(Filter(function(letter) (grepl(letter, "aiueoAIUEO")), strsplit(aString, split = "")[[1]]))
cat(paste(sep = "", countVowelsV7("Hello World"), "\n")) # 3

countVowelsV8 <- function(aString) length(Filter(function(letter) (grepl(toupper(letter), "AIUEO")), strsplit(aString, split = "")[[1]]))
cat(paste(sep = "", countVowelsV8("Hello World"), "\n")) # 3

File: codes/ruby/www.codewars/www.002.count_vowels.rb
=begin
Source:
    https://www.codewars.com/kata/54ff3102c1bad923760001f3
Title:
    Vowel Count
Description:
    Return the number (count) of vowels in the given string.
    We will consider a, e, i, o, u as vowels for this Kata (but not y).
    The input string will only consist of lower case letters and/or spaces.
=end

def count_vowels_v1(a_string)
    return a_string.split("").reduce(0) { |result, letter| (["a", "i", "u", "e", "o", "A", "I", "U", "E", "O"].include?(letter) ? (result + 1) : result) }
end
print(count_vowels_v1("Hello World"), "\n") # 3

def count_vowels_v2(a_string)
    return a_string.split("").reduce(0) { |result, letter| (["A", "I", "U", "E", "O"].include?(letter.upcase()) ? (result + 1) : result) }
end
print(count_vowels_v2("Hello World"), "\n") # 3

def count_vowels_v3(a_string)
    return a_string.split("").reduce(0) { |result, letter| ("aiueoAIUEO".include?(letter) ? (result + 1) : result) }
end
print(count_vowels_v3("Hello World"), "\n") # 3

def count_vowels_v4(a_string)
    return a_string.split("").reduce(0) { |result, letter| ("AIUEO".include?(letter.upcase()) ? (result + 1) : result) }
end
print(count_vowels_v4("Hello World"), "\n") # 3

def count_vowels_v5(a_string)
    return a_string.split("").select { |letter| ["a", "i", "u", "e", "o", "A", "I", "U", "E", "O"].include?(letter) }.size()
end
print(count_vowels_v5("Hello World"), "\n") # 3

def count_vowels_v6(a_string)
    return a_string.split("").select { |letter| ["A", "I", "U", "E", "O"].include?(letter.upcase()) }.size()
end
print(count_vowels_v6("Hello World"), "\n") # 3

def count_vowels_v7(a_string)
    return a_string.split("").select { |letter| "aiueoAIUEO".include?(letter) }.size()
end
print(count_vowels_v7("Hello World"), "\n") # 3

def count_vowels_v8(a_string)
    return a_string.split("").select { |letter| "AIUEO".include?(letter.upcase()) }.size()
end
print(count_vowels_v8("Hello World"), "\n") # 3
